### PR TITLE
Refactor for reusing easily

### PR DESCRIPTION
**Pull request by yslib**
1. replace unnecessary std::function by generic type.
2. Decouple app logic from SDL window framework(except for SDL_Event structure)

Details are included in commit and code comment.

### CODE REVIEW

**Review by harry75369 — COMMENTED** [review]
Running against a sketch file shows incorrect rendering
1) no cursor shown
2) incorrect transform

File: src/Utils/App.hpp [hunk 1]
```diff
@@ -39,31 +39,86 @@
...
+ * void pollEvent():
+ *         It will be invoked in every frame to poll event.
+ *         dispatchEvent() and dispatchGlobalEvent() must be processed properly in pollEvent()
+ *
+ * SDL event struct is adopted as our event handler framework, sdl runtime is not necessary for App
+ *
  * NOTE:
  * It must be derived to use this class, and the singleton app
- * instance can be obtained by App::getIntance<DerivedApp>().
+ * instance can be obtained by App<DerivedApp>::getIntance<DerivedApp>().
```
[the remark this answers is not in the document]
Comment (yslib) [reply src/Utils/App.hpp:79]: The redundant template parameter has been removed. I would find a sketch file and test it later.

[the remark this answers is not in the document]
Comment (yslib) [reply src/Utils/App.hpp:79]: Viewport size is wrong. It was caused by a using a ratio before it is evaluated correctly.

[the remark this answers is not in the document]
Comment (harry75369) [reply src/Utils/App.hpp:79]: > The redundant template parameter has been removed. I would find a sketch file and test it later.

The usage in the comment is still not correct. Maybe `App<DerivedApp>::getIntance()` ?

File: src/runtime.cpp [hunk 2]
```diff
@@ -16,186 +16,227 @@
...
+    // get device pixel ratio
+    int dw = 0, dh = 0;
+    int ww = 0, wh = 0;
+    SDL_GL_GetDrawableSize(window, &dw, &dh);
+    SDL_GetWindowSize(window, &ww, &wh);
+    m_properties["viewport_w"] = dw;
+    m_properties["viewport_h"] = dh;
+    m_properties["window_w"] = ww;
+    m_properties["window_h"] = wh;
+    m_properties["app_w"] = w;
```
Comment (harry75369) [inline src/runtime.cpp:173]: Duplication with member data `m_width` and `m_height`. If we use the properties, we should turn other member data into properties as well. And properties need both getter and setter, like `getProperty` and `setProperty`. Currently only getter exists.

Or we just don't use properties if there are no actual benefits.

**Review by harry75369 — COMMENTED** [review]
Besides, the use of CRTP pattern could be improved further:

1. Add a inline function to do both `static_assert(std::is_base_of...)` and `static_cast` stuff to prevent code duplication.

2. Make the interface constraints type-safe. For now, if derived class didn't implement one of the interfaces(initContext/makeContextCurrent/getProperty/swapBuffer/onInit/pollEvent) mentioned in the comments, it still compiles but will get a segment fault when running. Such cases should better be prevented at compile time.

File: src/Utils/App.hpp [hunk 3]
```diff
@@ -39,31 +39,98 @@
...
+  };
+
+HAS_MEMBER_FUNCTION_DEF(initContext)
+HAS_MEMBER_FUNCTION_DEF(makeContextCurrent)
+HAS_MEMBER_FUNCTION_DEF(getProperty)
+HAS_MEMBER_FUNCTION_DEF(setProperty)
+HAS_MEMBER_FUNCTION_DEF(swapBuffer)
+HAS_MEMBER_FUNCTION_DEF(onInit)
+HAS_MEMBER_FUNCTION_DEF(pollEvent)
+
```
Comment (harry75369) [inline src/Utils/App.hpp:110]: Add `undef` if this macro is no longer used.

**Pull request merged**